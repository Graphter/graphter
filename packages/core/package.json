{
  "name": "@graphter/core",
  "version": "1.0.0",
  "description": "",
  "main": "dist/index.js",
  "module": "dist/index.es.js",
  "files": [
    "dist"
  ],
  "types": "dist/index.d.ts",
  "scripts": {
    "test": "jest --config jest.config.js",
    "test:watch": "react-scripts test --env=jsdom",
    "test:update-snapshots": "jest --updateSnapshot",
    "build": "rollup -c",
    "start": "rollup -c -w",
    "prepare": "npm run build"
  },
  "author": "Jonathan Pritchard",
  "license": "ISC",
  "peerDependencies": {
    "react": "^16.13.1",
    "recoil": "0.0.8"
  },
  "dependencies": {},
  "devDependencies": {
    "@babel/core": "^7.9.0",
    "@babel/plugin-proposal-export-default-from": "^7.8.3",
    "@babel/preset-env": "^7.9.0",
    "@babel/preset-stage-0": "^7.8.3",
    "@rollup/plugin-json": "^4.0.3",
    "@types/jest": "^25.2.1",
    "@types/react": "^16.9.49",
    "@types/recoil": "0.0.1",
    "ajv": "^6.12.2",
    "eslint": "^6.8.0",
    "jest": "^25.5.4",
    "rollup": "^2.2.0",
    "rollup-plugin-babel": "^4.4.0",
    "rollup-plugin-commonjs": "^10.1.0",
    "rollup-plugin-json": "^4.0.0",
    "rollup-plugin-node-resolve": "^5.2.0",
    "rollup-plugin-peer-deps-external": "^2.2.2",
    "rollup-plugin-typescript2": "^0.26.0",
    "ts-jest": "^25.3.1",
    "typescript": "^3.8.3"
  }
}
