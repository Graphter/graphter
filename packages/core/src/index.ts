export * from './ListDisplayConfig'
export * from './Service'
export * from './ListResult'
export * from './GetResult'
export * from './SaveResult'
export * from './NodeConfig'
export * from './NodeRendererProps'
export * from './ErrorRendererProps'
export * from './NodeRendererRegistration'
export * from './JsonType'
export * from './PathSegment'
export * from './NodeValidation'
export * from './ValidationErrorDisplalyMode'
export * from './ValidationExecutionStage'
export * from './ValidationResult'
export * from './NodeValidatorRegistration'