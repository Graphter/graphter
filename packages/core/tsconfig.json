{
  "compilerOptions": {
    "declaration": true,
    "declarationDir": "./dist",
    "module": "esnext",
    "outDir": "./dist",
    "target": "es5",
    "jsx": "react",
    "noImplicitReturns": true,
    "noImplicitThis": true,
    "noImplicitAny": true,
    "strictNullChecks": true,
    "allowSyntheticDefaultImports": true,
    "moduleResolution": "node",
    "allowJs": true,
    "sourceMap": true,
    "lib": [
      "dom",
      "dom.iterable",
      "esnext"
    ],
    "skipLibCheck": true,
    "esModuleInterop": true,
    "strict": true,
    "forceConsistentCasingInFileNames": true,
    "resolveJsonModule": true,
    "isolatedModules": true,
    "noEmit": true
  },
  "include": [
    "src/**/*"
  ],
  "exclude": [
    "node_modules"
  ]
}
