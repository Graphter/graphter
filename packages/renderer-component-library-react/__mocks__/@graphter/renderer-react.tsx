import React, { useState } from "react";

export const useArrayNodeData = jest.fn(() => {
  return {
    childIds: [],
    removeItem: jest.fn(),
    commitItem: jest.fn()
  }
})

export const useNodeData = jest.fn(() => {
  return [
    null,
    () => {}
  ]
})

export const useNodeValidation = jest.fn().mockResolvedValue([])

export const nodeRendererStore = {
  get: jest.fn((nodeType: string) => ({
    type: 'string',
    renderer: ({originalNodeData}: any) => {
      const [ value, setValue ] = useState(originalNodeData || '')
      return (
        <input value={value} onChange={e => setValue(e.target.value)} data-testid='string-renderer' />
    )
    }
  }))
}
export const createDefault = jest.fn()