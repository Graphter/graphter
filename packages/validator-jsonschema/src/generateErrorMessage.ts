import { JsonSchemaValidatorOptions } from "./JsonSchemaValidatorOptions";
import { ValidateFunction } from "ajv";

export default function(data: any, options: JsonSchemaValidatorOptions, validateFn: ValidateFunction){
  if(typeof options.error === 'function'){
    return options.error(data);
  }
  if(typeof options.error === 'string'){
    return options.error;
  }
  if(validateFn.errors) {
    return validateFn.errors.map(err => err.message).join('. ');
  }
  return 'Some validation rule failed';
}