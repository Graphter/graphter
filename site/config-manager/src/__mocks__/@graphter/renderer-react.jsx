"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.pathConfigStore = exports.createDefault = exports.nodeRendererStore = exports.useNodeValidation = exports.useNodeData = exports.useArrayNodeData = void 0;
var react_1 = require("react");
exports.useArrayNodeData = jest.fn(function () {
    return {
        childIds: [],
        removeItem: jest.fn(),
        commitItem: jest.fn()
    };
});
exports.useNodeData = jest.fn(function () {
    return [
        null,
        function () { }
    ];
});
exports.useNodeValidation = jest.fn().mockResolvedValue([]);
exports.nodeRendererStore = {
    get: jest.fn(function (nodeType) { return ({
        type: 'string',
        renderer: function (_a) {
            var originalNodeData = _a.originalNodeData;
            var _b = react_1.useState(originalNodeData || ''), value = _b[0], setValue = _b[1];
            return (<input value={value} onChange={function (e) { return setValue(e.target.value); }} data-testid='string-renderer'/>);
        }
    }); })
};
exports.createDefault = jest.fn();
exports.pathConfigStore = {
    get: jest.fn(),
    set: jest.fn()
};
