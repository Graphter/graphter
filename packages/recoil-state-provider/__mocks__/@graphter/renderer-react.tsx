import { NodeRendererRegistration } from "@graphter/core";
import React from "react";

export const nodeRendererStore = {
  get: jest.fn()
}

export const pathConfigStore = {
  get: jest.fn(),
  set: jest.fn()
}

export const NodeValidationProvider = jest.fn(({ children }: any) => <div>{children}</div>)
export const NodeDataProvider = jest.fn(({ children }: any) => <div>{children}</div>)