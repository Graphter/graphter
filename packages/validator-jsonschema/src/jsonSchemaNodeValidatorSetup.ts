import { JsonSchemaValidatorOptions } from "./JsonSchemaValidatorOptions";
import {
  NodeConfig, ValidationErrorDisplayMode,
  ValidationExecutionStage,
  ValidationResult
} from "@graphter/core";
import generateErrorMessage from "./generateErrorMessage";
import Ajv from "ajv";

export default function jsonSchemaNodeValidatorSetup(options: JsonSchemaValidatorOptions){
  if(!options || !options.schema) throw new Error(`Schema option is required by JsonSchemaValidator`);
  options = {
    executeOn: [
      ValidationExecutionStage.CLIENT_CREATE,
      ValidationExecutionStage.CLIENT_UPDATE,
      ValidationExecutionStage.SERVER_UPDATE,
      ValidationExecutionStage.SERVER_UPDATE
    ],
    errorDisplayMode: [
      ValidationErrorDisplayMode.INLINE,
      ValidationErrorDisplayMode.SUMMARY
    ],
    ...options
  };

  if(options.executeOn === undefined) throw new Error(`Shouldn't happen if defaulting was done properly`)

  const ajv = new Ajv({
    jsonPointers: true,
    allErrors: true
  });
  const validateFn = ajv.compile(options.schema);


  return function execute(
    stage: ValidationExecutionStage,
    config: NodeConfig,
    data: any,
  ): Promise<ValidationResult> {
    const valid = validateFn(data);
    console.log(`${config.id} value of ${JSON.stringify(data)} ${valid?'passed':'failed'} validation: ${JSON.stringify(options)}`)
    if(valid){
      return Promise.resolve({
        valid: true,
        dataPaths: [ config.id ]
      });
    } else {
      const result: ValidationResult = {
        valid: false,
        errorMessage: generateErrorMessage(data, options, validateFn),
      }

      if(validateFn.errors){
        result.dataPaths = validateFn.errors.map(err => config.id + err.dataPath);
      }

      if(Array.isArray(options.errorDisplayMode)){
        result.errorDisplayMode = options.errorDisplayMode;
      } else if(options.errorDisplayMode !== undefined){
        result.errorDisplayMode = [ options.errorDisplayMode ];
      }

      return Promise.resolve(result);
    }
  }

}