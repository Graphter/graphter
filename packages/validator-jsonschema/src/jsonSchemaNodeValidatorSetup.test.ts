import AJV from 'ajv'
import generateErrorMessage from "./generateErrorMessage";

import jsonSchemaNodeValidatorSetup from './jsonSchemaNodeValidatorSetup'
import { ValidationExecutionStage } from "@graphter/core";
import ajvKeywords from "ajv-keywords";
import Ajv from "ajv";

jest.mock('ajv', () => {
  const MockAjv = {
    compile: jest.fn()
  }
  return jest.fn(() => MockAjv)
})
jest.mock('ajv-keywords')
jest.mock('./generateErrorMessage')

const AJVMock = AJV as jest.Mocked<any>
const generateErrorMessageMock = generateErrorMessage as jest.Mock<any>

describe('jsonSchemaNodeValidatorSetup', () => {
  let ajvMock: jest.Mocked<any>
  beforeEach(() => {
    ajvMock =  new Ajv();
  })

  it('should compile the schema to a validator function ', async () => {
    ajvMock.compile.mockReturnValueOnce(() => true)
    await jsonSchemaNodeValidatorSetup({ schema: { some: 'schema' } })
    expect(ajvKeywords).toHaveBeenCalledWith(ajvMock)
  })

  it('should use ajv keywords library', async () => {
    ajvMock.compile.mockReturnValueOnce(() => true)
    await jsonSchemaNodeValidatorSetup({ schema: { some: 'schema' } })
    expect(ajvMock.compile).toHaveBeenCalledWith({ some: 'schema' })
  })

  it('should use sensible defaults', async () => {
    ajvMock.compile.mockReturnValueOnce(() => false)
    const execute = await jsonSchemaNodeValidatorSetup({ schema: {} })
    await execute(ValidationExecutionStage.CHANGE,
      { id: 'name', type: 'string' },
      ['page'],
      'some-data')
    expect(generateErrorMessageMock).toHaveBeenCalled()
    expect(generateErrorMessageMock.mock.calls[0][1]).toMatchSnapshot()
  })

  it('should execute the validator function', async () => {
    const validatorMock = jest.fn().mockReturnValueOnce(true)
    ajvMock.compile.mockReturnValueOnce(validatorMock)
    const execute = await jsonSchemaNodeValidatorSetup({ schema: {} })
    await execute(ValidationExecutionStage.CHANGE,
      { id: 'name', type: 'string' },
      ['page'],
      'some-data')
    expect(validatorMock).toHaveBeenCalledWith('some-data')
  })

  it('should return the correct result when data is valid', async () => {
    const validatorMock = jest.fn().mockReturnValueOnce(true)
    ajvMock.compile.mockReturnValueOnce(validatorMock)
    const execute = await jsonSchemaNodeValidatorSetup({ schema: {} })
    const result = await execute(ValidationExecutionStage.CHANGE,
      { id: 'name', type: 'string' },
      ['page'],
      'some-data')
    expect(result).toMatchSnapshot()
  })

  it('should return the correct result when data is not valid', async () => {
    const validatorMock = jest.fn().mockReturnValueOnce(false)
    ajvMock.compile.mockReturnValueOnce(validatorMock)
    const execute = await jsonSchemaNodeValidatorSetup({ schema: {} })
    const result = await execute(ValidationExecutionStage.CHANGE,
      { id: 'name', type: 'string' },
      ['page'],
      'some-data')
    expect(result).toMatchSnapshot()
  })

  it('should error when no options are provided', async () => {
    // @ts-ignore
    await expect(() => jsonSchemaNodeValidatorSetup())
      .toThrow('Schema option is required by JsonSchemaValidator')
  })

  it('should error when no schema is provided', async () => {
    // @ts-ignore
    await expect(() => jsonSchemaNodeValidatorSetup({}))
      .toThrow('Schema option is required by JsonSchemaValidator')
  })


})

export {}