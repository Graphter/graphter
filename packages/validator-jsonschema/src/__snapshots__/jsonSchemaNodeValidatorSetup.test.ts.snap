// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`jsonSchemaNodeValidatorSetup should return the correct result when data is not valid 1`] = `
Object {
  "errorDisplayMode": Array [
    "INLINE",
    "SUMMARY",
  ],
  "errorMessage": undefined,
  "valid": false,
}
`;

exports[`jsonSchemaNodeValidatorSetup should return the correct result when data is valid 1`] = `
Object {
  "valid": true,
}
`;

exports[`jsonSchemaNodeValidatorSetup should use sensible defaults 1`] = `
Object {
  "errorDisplayMode": Array [
    "INLINE",
    "SUMMARY",
  ],
  "schema": Object {},
}
`;
